### PR TITLE
[feature] Docker

### PR DESCRIPTION
**Pull request by wnuke**
Add Dockerfile for using this as a docker container.

### CODE REVIEW

**Review by dominikaaaa — CHANGES_REQUESTED** [review]
Can you put the shell scripts in the scripts folder?

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+COPY src src
+
+RUN gradle build
+
+FROM openjdk:14
+
+WORKDIR /bot
+
+COPY --from=build /build/build/libs/bot-kt-*.jar /bot/bot-kt.jar
```
Comment (dominikaaaa) [inline Dockerfile:18]: This will keep old versions. Does docker redownload the Gradle cache every time or does it use a fresh clone?

Comment (wnuke) [reply Dockerfile:18]: If the contents of the first container changes, then the second container will be remade, everything downstream of a change gets rebuilt.

Comment (dominikaaaa) [reply Dockerfile:18]: Yes, but does it delete the build folder or use a fresh clone?

Comment (wnuke) [reply Dockerfile:18]: Anything that gets created on that step would be removed, so if step one is copy build dependency definition, step two is download dependencies, step three is copy source, step four build, etc... and source changes, then step 3 (and everything after/that depends on that step) gets rebuilt.

**Comment by dominikaaaa** [comment]
thanks. I'll test this later

**Comment by wnuke** [comment]
Cool, no problem.

**Review by sourTaste000 — APPROVED** [review]
lgtm, but can you make it upgrade the submodules everytime it builds?  
use this command
`git submodule foreach git pull origin master`

**Comment by l1ving** [comment]
The repository should be in charge of that

**Comment by l1ving** [comment]
You don't want the docker building different code then a local clone would build

**Review by l1ving — CHANGES_REQUESTED** [review]
Sorry again for the requested change, but I'd prefer if this was more inline with our current release system. Right now the bot will download the latest release and then restart itself if it's enabled in the config, code is [here](https://github.com/kami-blue/bot-kt/blob/master/src/main/kotlin/UpdateHelper.kt)

So all it would have to do is run the jar file from somewhere the bot can modify and restart if it crashes. If that's too much work I can merge it as is but I'd probably end up having to write my own docker image for our use case.

**Comment by wnuke** [comment]
> lgtm, but can you make it upgrade the submodules everytime it builds?  
> use this command
> `git submodule foreach git pull origin master`

As living said  this should be handled before docker builds by the repo/builder.

**Comment by wnuke** [comment]
> Sorry again for the requested change, but I'd prefer if this was more inline with our current release system. Right now the bot will download the latest release and then restart itself if it's enabled in the config, code is [here](https://github.com/kami-blue/bot-kt/blob/master/src/main/kotlin/UpdateHelper.kt)
> 
> So all it would have to do is run the jar file from somewhere the bot can modify and restart if it crashes. If that's too much work I can merge it as is but I'd probably end up having to write my own docker image for our use case.

The bot can modify anything in the container, that's why you would use the container, the bot can modify whatever it wants without affecting the host, which means that you will only update the bot in that container where the auto updater runs. I highly recommend detecting changes in the git repo and rebuilding the Docker image instead, but the current auto updater should work.

**Review by l1ving — APPROVED** [review]
Looks good to me, I'll test this tomorrow. Thanks

**Comment by sourTaste000** [comment]
tfw failed check

**Comment by sourTaste000** [comment]
check [this](https://www.codefactor.io/repository/github/kami-blue/bot-kt/pull/36)?

**Comment by l1ving** [comment]
Thanks, I'll test in a bit

**Comment by wnuke** [comment]
No problem!

File: .github/workflows/docker-publish.yml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+on:
+  push:
+    # Publish `master` as Docker `latest` image.
+    branches:
+      - master
+
+    # Publish `v1.2.3` tags as releases.
+    tags:
+      - v*
```
Comment (l1ving) [inline .github/workflows/docker-publish.yml:11]: I just realized, I don't use a v prefix for versioning in bot-kt, can you fix this?

File: .github/workflows/docker-publish.yml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+  # Run tests for any PRs.
+  pull_request:
```
Comment (l1ving) [inline .github/workflows/docker-publish.yml:14]: Is this meant to be empty? I'd like docker to run on PRs as well

File: .github/workflows/docker-publish.yml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+          IMAGE_ID=ghcr.io/${{ github.repository_owner }}/$IMAGE_NAME
+
+          # Change all uppercase to lowercase
+          IMAGE_ID=$(echo $IMAGE_ID | tr '[A-Z]' '[a-z]')
+
+          # Strip git ref prefix from version
+          VERSION=$(echo "${{ github.ref }}" | sed -e 's,.*/\(.*\),\1,')
+
+          # Strip "v" prefix from tag name
+          [[ "${{ github.ref }}" == "refs/tags/"* ]] && VERSION=$(echo $VERSION | sed -e 's/^v//')
```
Comment (l1ving) [inline .github/workflows/docker-publish.yml:70]: unnecessary, see above

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+RUN gradle build
+
+FROM openjdk:14
+
+WORKDIR /bot
+
+COPY --from=build /build/build/libs/bot-kt-*.jar /bot/
+
+RUN string1="$(ls bot-kt-*.jar)" && string2=${string1#"bot-kt-"} && echo ${string2%".jar"} > currentVersion
```
Comment (l1ving) [inline Dockerfile:20]: does currentVersion exist in docker builds? it's a file that's autogenerated when the bot is run...

**Review by l1ving — CHANGES_REQUESTED** [review]

File: .github/workflows/docker-publish.yml (hunk 3 above)
Comment (wnuke) [reply .github/workflows/docker-publish.yml:14]: It appears to be running the tests on this PR, I cannot remember why thats there, I just took the default docker publish action

Comment (l1ving) [reply .github/workflows/docker-publish.yml:14]: Okay, I guess it works if it's empty lol

File: Dockerfile (hunk 5 above)
Comment (wnuke) [reply Dockerfile:20]: Current version is created by the Dockerfile (when docker build is run)

File: .github/workflows/docker-publish.yml (hunk 2 above)
Comment (wnuke) [reply .github/workflows/docker-publish.yml:11]: Yes, of course

File: Dockerfile (hunk 5 above)
Comment (l1ving) [reply Dockerfile:20]: Sounds good

File: .github/workflows/docker-publish.yml (hunk 3 above)
Comment (wnuke) [reply .github/workflows/docker-publish.yml:14]: It probably is supposed to be empty unless you want it to publish, but again, not sure.

File: .github/workflows/docker-publish.yml (hunk 2 above)
Comment (wnuke) [reply .github/workflows/docker-publish.yml:11]: Appears to stop this from running on PRs...

**Review by l1ving — APPROVED** [review]

**Pull request merged**